[VPlan] Fix LastActiveLane assertion on scalar VF

lukel97 opened the pull request:
For a scalar only VPlan with tail folding, if it has a phi live out then legalizeAndOptimizeInductions will scalarize the widened canonical IV feeding into the header mask:

    <x1> vector loop: {
      vector.body:
        EMIT vp<%4> = CANONICAL-INDUCTION ir<0>, vp<%index.next>
        vp<%5> = SCALAR-STEPS vp<%4>, ir<1>, vp<%0>
        EMIT vp<%6> = icmp ule vp<%5>, vp<%3>
        EMIT vp<%index.next> = add nuw vp<%4>, vp<%1>
        EMIT branch-on-count vp<%index.next>, vp<%2>
      No successors
    }
    Successor(s): middle.block

    middle.block:
      EMIT vp<%8> = last-active-lane vp<%6>
      EMIT vp<%9> = extract-lane vp<%8>, vp<%5>
    Successor(s): ir-bb<exit>

The verifier complains about this but this should still generate the correct last active lane, so this fixes the assert by handling this case in isHeaderMask. There is a similar pattern already there for ActiveLaneMask, which also expects a VPScalarIVSteps recipe.

Fixes #167813

lukel97 committed:
[VPlan] Fix LastActiveLane assertion on scalar VF

For a scalar only VPlan with tail folding, if it has a phi live out then legalizeAndOptimizeInductions will scalarize the widened canonical IV feeding into the header mask:

    <x1> vector loop: {
      vector.body:
        EMIT vp<%4> = CANONICAL-INDUCTION ir<0>, vp<%index.next>
        vp<%5> = SCALAR-STEPS vp<%4>, ir<1>, vp<%0>
        EMIT vp<%6> = icmp ule vp<%5>, vp<%3>
        EMIT vp<%index.next> = add nuw vp<%4>, vp<%1>
        EMIT branch-on-count vp<%index.next>, vp<%2>
      No successors
    }
    Successor(s): middle.block

    middle.block:
      EMIT vp<%8> = last-active-lane vp<%6>
      EMIT vp<%9> = extract-lane vp<%8>, vp<%5>
    Successor(s): ir-bb<exit>

The verifier complains about this but this should still generate the correct last active lane, so this fixes the assert by handling this case in isHeaderMask. There is a similar pattern already there for ActiveLaneMask, which also expects a VPScalarIVSteps recipe.

Fixes #167813

lukel97 committed:
Remove trivial branch

lukel97 committed:
Add comment and assert

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/fix-tail-folded-live-out-scalar-vf

lukel97 committed:
Update test after merging

lukel97 commented:
I see that #149042 was reverted but I've updated this on main if you'd like to land this separately anyway.

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanUtils.cpp:
```diff
@@ -66,6 +66,17 @@ bool vputils::isHeaderMask(const VPValue *V, const VPlan &Plan) {
 
+  // For scalar plans, the header mask uses the scalar steps.
+  if (match(V,
+            m_ICmp(m_ScalarIVSteps(
```
Looks like this pattern is now used twice in this function, i.e.

```
m_ScalarIVSteps(m_Specific(Plan.getVectorLoopRegion()->getCanonicalIV()),
                m_One(), m_Specific(&Plan.getVF()))
```

Is it worth creating a specific pattern match for this? Something like

```
  m_CanonicalScalarIVSteps(Plan)
```

?

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/fix-tail-folded-live-out-scalar-vf

lukel97 committed:
Add m_CanonicalScalarIVSteps helper

arcbbb reviewed (commented):
Looks good to me. Thanks for the fix!

lukel97 committed:
Move to variable

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanUtils.cpp:
```diff
@@ -46,6 +46,12 @@ VPValue *vputils::getOrCreateVPValueForSCEVExpr(VPlan &Plan, const SCEV *Expr) {
   return Expanded;
 }
 
+static inline auto m_CanonicalScalarIVSteps(const VPlan &Plan) {
```
Yup, turned it into a variable in c0a70672c2a9f64b4efcc71ee786eb2234f9261e

fhahn commented on llvm/test/Transforms/LoopVectorize/tail-folding-live-out-scalar-vf.ll:
```diff
@@ -0,0 +1,53 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --check-globals none --version 6
+; RUN: opt -p loop-vectorize -prefer-predicate-over-epilogue=predicate-else-scalar-epilogue -force-vector-width=1 -force-vector-interleave=2 -S %s | FileCheck %s
+
+define i64 @live_out_scalar_vf(i64 %n) {
```
if possible, might be good to add this to `llvm/test/Transforms/LoopVectorize/tail-folding-vectorization-factor-1.ll`, which already contains a set of test cases for interleaving only with tail-folding.

fhahn reviewed (approved):
LGTM, thanks

lukel97 committed:
Move test to tail-folding-vectorization-factor-1.ll

Needed to add -prefer-predicate-over-epilogue=predicate-else-scalar-epilogue flag to run lines. Looks like the existing tests weren't necessarily tail folded?
